src/olympia/blocklist/admin.py

```diff
@@ -229,33 +235,36 @@ def delete_model(self, request, obj):
         super().delete_model(request, obj)
         ActivityLog.create(*args)
 
-    def _get_addon_versions(self):
+    def _get_addon_versions(self, obj):
         """Add some caching on the version queries.
-        We add it to the addon_instance object rather than self because the
+        We add it to the object rather than self because the
         ModelAdmin instance can be reused by subsequent requests.
         """
-        if not self.addon_instance:
+        if not obj or not obj.addon:
             return {}
-        elif not hasattr(self.addon_instance, '_addon_versions_cache'):
-            qs = self.addon_instance.versions(
+        elif not hasattr(obj, '_addon_versions_cache'):
+            qs = obj.addon.versions(
                 manager='unfiltered_for_relations').values(
                 'version', 'channel')
-            self.addon_instance._addon_versions_cache = {
+            obj._addon_versions_cache = {
                 version['version']: version['channel'] for version in qs}
-        return self.addon_instance._addon_versions_cache
+        return obj._addon_versions_cache
+
+    def get_request_guid(self, request):
+        return request.GET.get('guid')
 
     def get_form(self, request, obj=None, **kwargs):
-        if obj:
-            self.addon_instance = obj.addon
-        else:
-            self._guid = request.GET.get('guid')
-            self.addon_instance = Block(guid=self._guid).addon
+        # Grab the obj here so we can reference it in formfield_for_db_field,
+        # or create a new Block instance otherwise.
+        request._obj = (
+            Block(guid=self.get_request_guid(request)) if not obj else obj)
```

Comment:
I wonder if we could set the choices in this method (below, after `super().get_form(...)` to avoid setting `_obj` on `request` ?

Reply:
After a few hours of looking through the spaghetti soup of code in django admin model code, django forms, and django fields, I don't have an answer to that.  It certainly _seems_ like that should be possible, if one discovers the right method to hook into.

Reply:
https://github.com/mozilla/addons-server/pull/12832/commits/7d3f0b97434bc60ca00cfa86962ed39284dbe1b2

Reply:
My idea was to set the choices without using `formfield_for_dbfield` at all - letting django create the form (calling `super().get_form(...)`) and *then* modify the choices in the returned `form`. Is that not possible ?

Reply:
like I was saying, it seems like it should be possible, if one finds the right way of doing it.

Reply:
it's a question of how many (more) hours is to be spent trying to get it to work

Reply:
We need some way of forcing the field to be a `ChoiceField`, and setting the choices.  Forcing the field to have choices in the model causes django migration problems (and it's difficult to get the correct choices at the correct point when they depend on a query based on the value of another field).  I tried a lot of variations, but at some point you have to stop (and I was already double the time limit for the timebox I set myself)

Reply:
My naive implementation : https://gist.github.com/diox/5cd0f93cfcc14f9dc4d7e5b20bb7c3c0

It seems to work locally and pass the tests - but maybe more tests are needed.

Reply:
I had a thought while I was eating just now that something like that might work (if I could remember what the fields property was) - thanks for doing the legwork!